File: libs/qec/unittests/decoders/chromobius/test_chromobius.cpp
/*******************************************************************************
 * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

#include "cudaq/qec/decoder.h"

#include <gtest/gtest.h>
#include <string>
#include <string_view>
#include <vector>

namespace {

constexpr const char *chromobius_dem = R"DEM(
error(0.1000000000000000056) D0 D1
error(0.1000000000000000056) D0 D1 D2
error(0.1000000000000000056) D0 L0
error(0.1000000000000000056) D1 D2 D3
error(0.1000000000000000056) D2 D3
error(0.1000000000000000056) D3
detector(0, 0, 0, 1) D0
detector(1, 0, 0, 2) D1
detector(2, 0, 0, 0) D2
detector(3, 0, 0, 1) D3
)DEM";

cudaqx::tensor<uint8_t> make_H(std::size_t num_detectors = 4,
                               std::size_t num_observables = 1) {
  std::vector<uint8_t> H_data(num_detectors * num_observables, 0);
  cudaqx::tensor<uint8_t> H;
  H.copy(H_data.data(), {num_detectors, num_observables});
  return H;
}

cudaqx::heterogeneous_map make_params() {
  return cudaqx::heterogeneous_map{};
}

} // namespace

TEST(ChromobiusDecoder, checkAllZeroSyndrome) {
  auto decoder = cudaq::qec::decoder::get(
      "chromobius", std::string_view{chromobius_dem}, make_params());

  std::vector<cudaq::qec::float_t> syndrome = {0, 0, 0, 0};
  auto result = decoder->decode(syndrome);

  EXPECT_TRUE(result.converged);
  ASSERT_EQ(result.result.size(), 1);
  EXPECT_EQ(result.result[0], 0.0);
  EXPECT_EQ(decoder->get_block_size(), 1);
  EXPECT_EQ(decoder->get_syndrome_size(), 4);
}

TEST(ChromobiusDecoder, checkKnownObservableFlip) {
  auto decoder = cudaq::qec::decoder::get(
      "chromobius", std::string_view{chromobius_dem}, make_params());

  std::vector<cudaq::qec::float_t> syndrome = {1, 0, 0, 0};
  auto result = decoder->decode(syndrome);

  EXPECT_TRUE(result.converged);
  ASSERT_EQ(result.result.size(), 1);
  EXPECT_EQ(result.result[0], 1.0);
}

TEST(ChromobiusDecoder, checkPcmInputThrows) {
  EXPECT_THROW(
      (void)cudaq::qec::decoder::get("chromobius", make_H(), make_params()),
      std::runtime_error);
}

TEST(ChromobiusDecoder, checkMalformedDemThrows) {
  EXPECT_THROW((void)cudaq::qec::decoder::get(
                   "chromobius", std::string_view{"not a valid DEM"},
                   make_params()),
               std::runtime_error);
}

File: libs/qec/python/tests/test_chromobius.py
# ============================================================================ #
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

from pathlib import Path

import cudaq_qec as qec


def _reference_dir() -> Path:
    return Path(__file__).resolve().parents[2] / "test_data" / "chromobius"


def _read_dem() -> str:
    return (_reference_dir() /
            "basic_reference.dem").read_text(encoding="utf-8")


def _parse_bits(bits: str) -> list[float]:
    values = []
    for bit in bits.strip():
        if bit not in "01":
            raise ValueError(f"Invalid reference bit: {bit!r}")
        values.append(float(bit == "1"))
    return values


def _read_shots() -> list[tuple[str, list[float], list[float]]]:
    shots = []
    path = _reference_dir() / "basic_reference.tsv"
    for line_no, line in enumerate(
            path.read_text(encoding="utf-8").splitlines(), start=1):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        columns = line.split()
        if len(columns) != 3:
            raise ValueError(
                f"Malformed reference row {path}:{line_no}: {line}")
        name, syndrome_bits, expected_bits = columns
        shots.append(
            (name, _parse_bits(syndrome_bits), _parse_bits(expected_bits)))
    return shots


def _as_list(values):
    return values.tolist() if hasattr(values, "tolist") else list(values)


def _make_decoder():
    dem = _read_dem()
    shots = _read_shots()
    return qec.get_decoder("chromobius", dem), shots


def test_chromobius_matches_upstream_predict_reference():
    decoder, shots = _make_decoder()
    for name, syndrome, expected in shots:
        result = decoder.decode(syndrome)
        assert result.converged, name
        actual = _as_list(result.result)
        assert actual == expected, (name, actual, expected)


def test_chromobius_batch_matches_upstream_predict_reference():
    decoder, shots = _make_decoder()
    batch_results = decoder.decode_batch([syndrome for _, syndrome, _ in shots])
    for batch_result, (name, _, expected) in zip(batch_results, shots):
        assert batch_result.converged, name
        actual = _as_list(batch_result.result)
        assert actual == expected, (name, actual, expected)
